lib/ddtrace/configuration/settings.rb

```diff
@@ -82,7 +82,8 @@ class Settings
       end
 
       option :tracer do |o|
-        o.default Tracer.new
+        o.default { Tracer.new }
```

Comment:
Making the `RulerSampler` the default sampler creates a circular dependency:
1. Load this `Settings` class.
2. Eager load the tracer on this line: `o.default Tracer.new`
3. `Tracer` initializer create the default sampler, which is not the `RulerSampler`.
4. `RuleSampler` tries to read a setting from `Settings`: https://github.com/DataDog/dd-trace-rb/blob/dbcac580b0bbf314901ded18be7143eaaa5265cb/lib/ddtrace/sampling/rule_sampler.rb#L28

Making the tracer lazy like this will work on a real application, because it only has a singleton tracer instance.

But in our tests, we create a new tracer (`get_test_tracer`) which sets its context to the current thread (when a new `Datadog::DefaultContextProvider.new` is initialized): https://github.com/DataDog/dd-trace-rb/blob/8a9fb5529bc1491de4963542ac95b413b3bacac5/lib/ddtrace/tracer.rb#L76
This is fine, but we also call `Datadog.tracer` in various places of our code, which then triggers this lazy `o.default { Tracer.new }` that then overrides the context of the current thread once again, "stealing" all spans from that point on.

It seems like we probably need to make sure that only one tracer exists at one time, as simply calling `Tracer.new` has side-effects on the current thread that can't be easily undone. Another option is to make `Datadog::DefaultContextProvider.new` **not** set the thread context immediately.

Do you have any thoughts on this @delner?